### PR TITLE
use a sym to ind dict

### PR DESCRIPTION
**Pull request by chriselrod**

### CODE REVIEW

**Comment by chriselrod** [comment]
PR*:
```julia
julia> @benchmark ModelingToolkit.SymbolicIndexingInterface.state_sym_to_index($seir_model, $I)  
BenchmarkTools.Trial: 8023 samples with 907 evaluations.
 Range (min … max):  125.115 ns …  20.456 μs  ┊ GC (min … max): 0.00% … 99.06%
 Time  (median):     130.808 ns               ┊ GC (median):    0.00%
 Time  (mean ± σ):   136.253 ns ± 276.440 ns  ┊ GC (mean ± σ):  3.14% ±  1.56%

              ▅▆█▆▄▁                                             
  ▁▁▂▂▃▃▂▂▂▃▄▇██████▇▅▄▄▅▆█▇▇▇▇▆▆▅▄▄▃▃▃▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▁▁▁ ▃
  125 ns           Histogram: frequency by time          143 ns <

 Memory estimate: 16 bytes, allocs estimate: 1.
```
Master:
```julia
julia> @benchmark ModelingToolkit.SymbolicIndexingInterface.state_sym_to_index($seir_model, $I)  
BenchmarkTools.Trial: 10000 samples with 10 evaluations.
 Range (min … max):  1.202 μs … 984.542 μs  ┊ GC (min … max):  0.00% … 99.26%
 Time  (median):     1.297 μs               ┊ GC (median):     0.00%
 Time  (mean ± σ):   1.758 μs ±  18.701 μs  ┊ GC (mean ± σ):  24.67% ±  2.43%

      ▂▃▅▇█▇▆▅▄▃▂▁                                             
  ▁▂▄▆█████████████▇▅▅▄▄▃▃▃▃▃▂▂▂▂▂▂▂▂▂▂▂▂▂▂▁▁▁▁▂▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
  1.2 μs          Histogram: frequency by time        1.69 μs <

 Memory estimate: 2.42 KiB, allocs estimate: 19.
```

*With these PRs, to speed up dict lookups:
https://github.com/YingboMa/Unityper.jl/commit/85e77bfae317a94085dd16624a319ad085f8b111
https://github.com/JuliaSymbolics/Symbolics.jl/pull/927
https://github.com/JuliaSymbolics/SymbolicUtils.jl/pull/533

File: src/systems/diffeqs/odesystem.jl [hunk 1]
```diff
@@ -140,13 +140,19 @@ struct ODESystem <: AbstractODESystem
...
+
     function ODESystem(tag, deqs, iv, dvs, ps, tspan, var_to_name, ctrls, observed, tgrad,
         jac, ctrl_jac, Wfact, Wfact_t, name, systems, defaults,
         torn_matching, connector_type, preface, cevents,
         devents, metadata = nothing, gui_metadata = nothing,
         tearing_state = nothing,
         substitutions = nothing, complete = false,
-        discrete_subsystems = nothing, unknown_states = nothing;
+        discrete_subsystems = nothing, unknown_states = nothing,
+        sym_to_index = Dict{Num, Int}();
```
Comment (YingboMa) [inline src/systems/diffeqs/odesystem.jl:155]:
```suggestion
        sym_to_index = Dict{Any, Int}();
```

File: src/systems/diffeqs/odesystem.jl [hunk 2]
```diff
@@ -25,7 +25,7 @@ eqs = [D(x) ~ σ*(y-x),
 struct ODESystem <: AbstractODESystem
     """
     tag: a tag for the system. If two systems have the same tag, then they are
-    structurally identical.
+    structurally ide, sym_to_stringntical.
```
Comment (YingboMa) [inline src/systems/diffeqs/odesystem.jl:28]: Deliberate?

Comment (chriselrod) [reply src/systems/diffeqs/odesystem.jl:28]: No

**Comment by ChrisRackauckas** [comment]
@AayushSabharwal something of the sort is done at completion? I thought some indexing stuff was added

**Comment by AayushSabharwal** [comment]
Yeah indexes are cached in `Dict`s now.

**Pull request closed without merging**